File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Bridge/Bridge.py
from abc import ABC,abstractmethod
class Shape(ABC):
    @abstractmethod
    def draw(self)->str:
        pass
class RedCircle(Shape):
  """Implementation for drawing a red circle"""
  def draw(self) -> str:
    return "Drawing a red circle"

class BlueSquare(Shape):
  """Implementation for drawing a blue square"""
  def draw(self) -> str:
    return "Drawing a blue square"

def create_shape(shape_type):
  """Factory function to create shapes"""
  if shape_type == "red-circle":
    return RedCircle()
  elif shape_type == "blue-square":
    return BlueSquare()
  else:
    raise ValueError(f"Unsupported shape type: {shape_type}")

class Drawing:
  """Abstraction for drawing shapes"""
  def __init__(self, shape: Shape):
    self.shape = shape

  def render(self) -> str:
    """Delegates drawing to the specific shape implementation"""
    return self.shape.draw()

# Usage
red_circle = create_shape("red-circle")
blue_square = create_shape("blue-square")

red_drawing = Drawing(red_circle)
print(red_drawing.render())  # Output: Drawing a red circle

blue_drawing = Drawing(blue_square)
print(blue_drawing.render())  # Output: Drawing a blue square

File: 2.Intermediate Level/6.Software-Design-Patterns/example/Structural/Bridge/Bridge.ts
// Abstraction (Drawing API)
interface Drawable {
  draw(): string; // Returns a description of the drawing
}

// Implementations (Drawing with different colors)
class RedCircle implements Drawable {
  draw(): string {
    return "Drawing a red circle";
  }
}

class BlueSquare implements Drawable {
  draw(): string {
    return "Drawing a blue square";
  }
}

// Error handling class for potential Shape creation issues
class ShapeCreationError extends Error {
  constructor(message: string) {
    super(message);
  }
}

// Factory function to encapsulate Shape creation (optional)
function createShape(shapeType: string): Drawable {
  switch (shapeType.toLowerCase()) {
    case "redcircle":
      return new RedCircle();
    case "bluesquare":
      return new BlueSquare();
    default:
      throw new ShapeCreationError(`Unsupported shape type: ${shapeType}`);
  }
}

// Abstraction (with reference to implementation)
class Drawing {
  private shape: Drawable;

  constructor(shape: Drawable) {
    this.shape = shape;
  }

  public render(): string {
    return this.shape.draw();
  }
}

// Usage
try {
  const redCircle = createShape("RedCircle"); // Factory can be used for consistency
  const blueSquare = createShape("BlueSquare");

  const redDrawing = new Drawing(redCircle);
  console.log(redDrawing.render()); // Output: Drawing a red circle

  const blueDrawing = new Drawing(blueSquare);
  console.log(blueDrawing.render()); // Output: Drawing a blue square
} catch (error) {
  if (error instanceof ShapeCreationError) {
    console.error("Error creating shape:", error.message);
  } else {
    console.error("Unexpected error:", error);
  }
}
